provider/azure.go

```diff
@@ -0,0 +1,456 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"fmt"
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/Azure/azure-sdk-for-go/arm/dns"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/to"
+
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+	"github.com/kubernetes-incubator/external-dns/plan"
+)
+
+const (
+	azureRecordTTL = 300
+)
+
+type config struct {
+	Cloud          string `json:"cloud" yaml:"cloud"`
+	TenantID       string `json:"tenantId" yaml:"tenantId"`
+	SubscriptionID string `json:"subscriptionId" yaml:"subscriptionId"`
+	ResourceGroup  string `json:"resourceGroup" yaml:"resourceGroup"`
+	Location       string `json:"location" yaml:"location"`
+	ClientID       string `json:"aadClientId" yaml:"aadClientId"`
+	ClientSecret   string `json:"aadClientSecret" yaml:"aadClientSecret"`
+}
+
+// ZonesClient is an interface of dns.ZoneClient that can be stubbed for testing.
+type ZonesClient interface {
+	ListByResourceGroup(resourceGroupName string, top *int32) (result dns.ZoneListResult, err error)
+	ListNextResults(lastResults dns.ZoneListResult) (result dns.ZoneListResult, err error)
+}
+
+// RecordsClient is an interface of dns.RecordClient that can be stubbed for testing.
+type RecordsClient interface {
+	ListByDNSZone(resourceGroupName string, zoneName string, top *int32) (result dns.RecordSetListResult, err error)
+	ListByDNSZoneNextResults(list dns.RecordSetListResult) (result dns.RecordSetListResult, err error)
+	Delete(resourceGroupName string, zoneName string, relativeRecordSetName string, recordType dns.RecordType, ifMatch string) (result autorest.Response, err error)
+	CreateOrUpdate(resourceGroupName string, zoneName string, relativeRecordSetName string, recordType dns.RecordType, parameters dns.RecordSet, ifMatch string, ifNoneMatch string) (result dns.RecordSet, err error)
+}
+
+// AzureProvider implements the DNS provider for Microsoft's Azure cloud platform.
+type AzureProvider struct {
+	domainFilter  string
+	dryRun        bool
+	resourceGroup string
+	zonesClient   ZonesClient
+	recordsClient RecordsClient
+}
+
+// NewAzureProvider creates a new Azure provider.
+//
+// Returns the provider or an error if a provider could not be created.
+func NewAzureProvider(configFile string, domainFilter string, resourceGroup string, dryRun bool) (Provider, error) {
+	if configFile == "" {
+		return nil, fmt.Errorf("the --azure-config-file option is required")
```

Comment:
I'll remove this check and move it to `validation.go` and make the error message a little less specific so that it doesn't apply only to the command line option.